File: OsmAnd-java/src/main/java/net/osmand/router/RoadSplitStructure.java
package net.osmand.router;

import java.util.ArrayList;
import java.util.List;

public class RoadSplitStructure {
	boolean keepLeft = false;
	boolean keepRight = false;
	boolean speak = false;
	private static final float TURN_SLIGHT_DEGREE = 5;

	List<AttachedRoadInfo> leftLanesInfo = new ArrayList<>();
	int leftLanes = 0;
	int leftMaxPrio = 0;
	int roadsOnLeft = 0;

	List<AttachedRoadInfo> rightLanesInfo = new ArrayList<>();
	int rightLanes = 0;
	int rightMaxPrio = 0;
	int roadsOnRight = 0;

	public boolean allAreStraight() {
		for (AttachedRoadInfo angle : leftLanesInfo) {
			if (Math.abs(angle.attachedAngle) > TURN_SLIGHT_DEGREE) {
				return false;
			}
		}
		for (AttachedRoadInfo angle : rightLanesInfo) {
			if (Math.abs(angle.attachedAngle) > TURN_SLIGHT_DEGREE) {
				return false;
			}
		}
		return true;
	}

	public static class AttachedRoadInfo {
		int[] parsedLanes;
		double attachedAngle;
		int lanes;
		int speakPriority;
		public boolean attachedOnTheRight;
		public int turnType;
	}
	
	public String roadsInfo(List<AttachedRoadInfo> attached) {
		String s = "";
		if (attached != null) {
			int i = 0;
			for (AttachedRoadInfo a : attached) {
				s += String.format("\n  Road %d angle: %.2f, lanes: %d, lanesInfo: %s", 
						++i, a.attachedAngle, a.lanes, a.parsedLanes == null ? "" :TurnType.lanesToString(a.parsedLanes));
			}
		}
		return s;
	}
	
	@Override
	public String toString() {
		return String.format(" Left lanes %d roads %d prioRoads %d: %s\n Right lanes %d roads %d prioRoads: %d %s", 
				leftLanes, roadsOnLeft, leftMaxPrio, roadsInfo(leftLanesInfo),
				rightLanes, roadsOnRight, rightMaxPrio, roadsInfo(rightLanesInfo));
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/routing/RoadSplitStructure.kt
package net.osmand.shared.routing

import kotlin.jvm.JvmField
import kotlin.math.abs

/**
 * What the roads leaving one intersection look like from the road being driven: how many of them go
 * off to each side, how many lanes they take with them, and how loudly each deserves to be spoken.
 *
 * This is what a "keep left" or "keep right" is decided from - a fork is only worth an instruction
 * when there is something to keep away from.
 */
class RoadSplitStructure {

	@JvmField var keepLeft: Boolean = false
	@JvmField var keepRight: Boolean = false
	@JvmField var speak: Boolean = false

	@JvmField var leftLanesInfo: MutableList<AttachedRoadInfo> = ArrayList()
	@JvmField var leftLanes: Int = 0
	@JvmField var leftMaxPrio: Int = 0
	@JvmField var roadsOnLeft: Int = 0

	@JvmField var rightLanesInfo: MutableList<AttachedRoadInfo> = ArrayList()
	@JvmField var rightLanes: Int = 0
	@JvmField var rightMaxPrio: Int = 0
	@JvmField var roadsOnRight: Int = 0

	/** True when nothing here actually turns: every road leaves within the slight-turn margin. */
	fun allAreStraight(): Boolean {
		for (angle in leftLanesInfo) {
			if (abs(angle.attachedAngle) > TURN_SLIGHT_DEGREE) {
				return false
			}
		}
		for (angle in rightLanesInfo) {
			if (abs(angle.attachedAngle) > TURN_SLIGHT_DEGREE) {
				return false
			}
		}
		return true
	}

	/** One road leaving the intersection, as the turn logic needs to see it. */
	class AttachedRoadInfo {
		@JvmField var parsedLanes: IntArray? = null
		@JvmField var attachedAngle: Double = 0.0
		@JvmField var lanes: Int = 0
		@JvmField var speakPriority: Int = 0
		@JvmField var attachedOnTheRight: Boolean = false
		@JvmField var turnType: Int = 0
	}

	companion object {
		/** java had this as a float 5, widened to a double at every comparison below. */
		private const val TURN_SLIGHT_DEGREE = 5.0
	}
}
